### DISCUSSION TITLE
Running Marp Cli in Heroku fails

### DISCUSSION THREAD

**Discussion by amir734jj** [marp-team/marp#198] [Q&A] [answered]
I am trying to run Marp CLI in Heroku and I get the following error. I already added chrome and chrome driver build packs. I appreciate any help or hint. Thank you.

![image](https://user-images.githubusercontent.com/4202918/138620542-1f3a23f1-7065-4b77-9502-5cf2dd022357.png)

I even added [puppeteer build pack as suggested here](https://github.com/marp-team/marp/discussions/165)

          [  INFO ] Converting 1 markdown...
           [  WARN ] Insecure local file accessing is enabled for conversion from
                     lab01/note.marp.md.
           [ ERROR ] Failed converting Markdown. (Protocol error
                     (Target.setDiscoverTargets): Target closed.)
           child_process.js:866
               throw err;
               ^
           
           Error: Command failed: marp lab01/note.marp.md --allow-local-files --pdf -o lab01/note.marp.pdf
           [  INFO ] Converting 1 markdown...
           [  WARN ] Insecure local file accessing is enabled for conversion from
                     lab01/note.marp.md.
           [ ERROR ] Failed converting Markdown. (Protocol error
                     (Target.setDiscoverTargets): Target closed.)
           
               at checkExecSyncError (child_process.js:790:11)
               at Object.execSync (child_process.js:863:15)
               at /tmp/build_8c780db8/scripts/build.js:8:21
               at Array.forEach (<anonymous>)
               at /tmp/build_8c780db8/scripts/build.js:5:9
               at f (/tmp/build_8c780db8/node_modules/once/once.js:25:25)
               at Glob.<anonymous> (/tmp/build_8c780db8/node_modules/glob/glob.js:148:7)
               at Glob.emit (events.js:400:28)
               at Glob._finish (/tmp/build_8c780db8/node_modules/glob/glob.js:194:8)
               at done (/tmp/build_8c780db8/node_modules/glob/glob.js:179:14) {
             status: 1,
             signal: null,
             output: [
               null,
               Buffer(0) [Uint8Array] [],
               Buffer(245) [Uint8Array] [
                  91,  32,  32,  73,  78,  70,  79,  32,  93,  32,  67, 111,
                 110, 118, 101, 114, 116, 105, 110, 103,  32,  49,  32, 109,
                  97, 114, 107, 100, 111, 119, 110,  46,  46,  46,  10,  91,
                  32,  32,  87,  65,  82,  78,  32,  93,  32,  73, 110, 115,
                 101,  99, 117, 114, 101,  32, 108, 111,  99,  97, 108,  32,
                 102, 105, 108, 101,  32,  97,  99,  99, 101, 115, 115, 105,
                 110, 103,  32, 105, 115,  32, 101, 110,  97,  98, 108, 101,
                 100,  32, 102, 111, 114,  32,  99, 111, 110, 118, 101, 114,
                 115, 105, 111, 110,
                 ... 145 more items
               ]
             ],
             pid: 1728,
             stdout: Buffer(0) [Uint8Array] [],
             stderr: Buffer(245) [Uint8Array] [
                91,  32,  32,  73,  78,  70,  79,  32,  93,  32,  67, 111,
               110, 118, 101, 114, 116, 105, 110, 103,  32,  49,  32, 109,
                97, 114, 107, 100, 111, 119, 110,  46,  46,  46,  10,  91,
                32,  32,  87,  65,  82,  78,  32,  93,  32,  73, 110, 115,
               101,  99, 117, 114, 101,  32, 108, 111,  99,  97, 108,  32,
               102, 105, 108, 101,  32,  97,  99,  99, 101, 115, 115, 105,
               110, 103,  32, 105, 115,  32, 101, 110,  97,  98, 108, 101,
               100,  32, 102, 111, 114,  32,  99, 111, 110, 118, 101, 114,
               115, 105, 111, 110,
               ... 145 more items
             ]
           }
           npm ERR! code ELIFECYCLE
           npm ERR! errno 1
           npm ERR! cs655-lecture-notes@1.0.0 build: `node ./scripts/build.js`
           npm ERR! Exit status 1
           npm ERR! 
           npm ERR! Failed at the cs655-lecture-notes@1.0.0 build script.
           npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
           
           npm ERR! A complete log of this run can be found in:
           npm ERR!     /tmp/npmcache.7HLKz/_logs/2021-10-24T22_47_04_090Z-debug.log
           npm ERR! code ELIFECYCLE
           npm ERR! errno 1
           npm ERR! cs655-lecture-notes@1.0.0 postinstall: `npm run build`
           npm ERR! Exit status 1
           npm ERR! 
           npm ERR! Failed at the cs655-lecture-notes@1.0.0 postinstall script.
           npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
           
           npm ERR! A complete log of this run can be found in:
           npm ERR!     /tmp/npmcache.7HLKz/_logs/2021-10-24T22_47_04_136Z-debug.log
    -----> Build failed
           
           We're sorry this build is failing! You can troubleshoot common issues here:
           https://devcenter.heroku.com/articles/troubleshooting-node-deploys
           
           Some possible problems:
           
           - Node version not specified in package.json
             https://devcenter.heroku.com/articles/nodejs-support#specifying-a-node-js-version
           
           Love,
           Heroku
           
     !     Push rejected, failed to compile Node.js app.
     !     Push failed

[Repository](https://github.com/amir734jj/cs655-lecture-notes/tree/1606d4b58a435a4e896e64f65101d8ad08bf09e6) and [script](https://github.com/amir734jj/cs655-lecture-notes/blob/1606d4b58a435a4e896e64f65101d8ad08bf09e6/scripts/build.js#L7)

**Comment by yhatt** [comment]
How about [setting `CHROME_PATH` env](https://github.com/marp-team/marp-cli#convert-to-pdf---pdf) to `/app/.apt/usr/bin/google-chrome` (or `$GOOGLE_CHROME_BIN` / `$GOOGLE_CHROME_SHIM`) while running Marp CLI?

> Additionally, chromedriver expects Chrome to be installed at /usr/bin/google-chrome, but that's a read-only filesystem in a Heroku slug. You'll need to tell Selenium/chromedriver that the chrome binary is at /app/.apt/usr/bin/google-chrome instead.
>
> _-- https://github.com/heroku/heroku-buildpack-google-chrome#selenium_

**Reply by amir734jj** [reply]
No success. I tried both. Getting the same error.

https://github.com/amir734jj/cs655-lecture-notes/blob/e16fa6697e40bf43a2e10fcf4f6599a6c22e3b53/scripts/build.js#L10

https://github.com/amir734jj/cs655-lecture-notes/blob/4b816ba602e725a79cdeb53962284bff21529341/scripts/build.js#L10

**Reply by yhatt** [reply]
Marp CLI recognizes `CHROME_PATH` env, not `CHROME_DRIVER` env.

**Reply by amir734jj** [reply]
No success. Getting the same error.

https://github.com/amir734jj/cs655-lecture-notes/commit/1d8902131f3f5faa1655919ee9c7f6e0db2c6e88

https://github.com/amir734jj/cs655-lecture-notes/commit/b572c3bc911c5f92482a89afbd67b73e636d31e6

**Reply by yhatt** [reply]
- Node.js's `exec()` might not parse `"$GOOGLE_CHROME_BIN"` because the spawning process has not run in the shell. Try to use `process.env.GOOGLE_CHROME_BIN` instead.
- If failed, try to set the actual executable path of Chrome: probably [`/app/.apt/opt/google/chrome/chrome`](https://github.com/heroku/heroku-buildpack-google-chrome/blob/6bb037f2aed3a49ee65c935f23cac1abe04847db/bin/compile#L192). Marp CLI cannot detect the path of Chrome installed by Heroku buildpack automatically, because the buildpack installs Chrome with a not-standard binary name `chrome`. [Here](https://github.com/GoogleChrome/chrome-launcher/blob/master/src/chrome-finder.ts#L108-L113) is how Marp CLI to find Chrome executable path.
- If failed again, try to set the path to the shim binary (`process.env.GOOGLE_CHROME_SHIM` or `/app/.apt/usr/bin/google-chrome-stable`). I have done using Heroku chrome buildpack in the other project (not Marp) before. At this time, I had succeeded to launch Chrome when changed to use the shim.

**Reply by amir734jj** [reply]
I tried all 3 and failed. I think I should give up :)

https://github.com/amir734jj/cs655-lecture-notes/blob/7cd737383451b5027135d78a1488826a5d36ae3b/scripts/build.js#L319


       > node ./scripts/build.js
       
       [
         '/tmp/build_1a1b8b77/.apt/usr/bin/google-chrome-stable',
         '/tmp/build_1a1b8b77/.apt/usr/bin/google-chrome'
       ]
       /tmp/build_1a1b8b77/node_modules/execa/index.js:212
       		throw error;
       		^
       
       Error: Command failed with exit code 1: marp lab01/note.marp.md --allow-local-files --pdf -o lab01/note.marp.pdf
       [  INFO ] Converting 1 markdown...
       [  WARN ] Insecure local file accessing is enabled for conversion from
                 lab01/note.marp.md.
       [ ERROR ] Failed converting Markdown. (Protocol error
                 (Target.setDiscoverTargets): Target closed.)

**Reply by yhatt** [reply]
Probably there is no choice but to stand against the root cause. I've created a new ticket at marp-team/marp-cli#395. Track it if interested.

**Comment by yhatt** [comment, marked as answer]
**UPDATED:** It may have resolved in [Marp CLI v1.4.2](https://github.com/marp-team/marp-cli/releases/tag/v1.4.2).

- https://github.com/marp-team/marp-cli/issues/395
- https://github.com/marp-team/marp-cli/pull/400

**Reply by amir734jj** [reply]
Yes. Thank you. It's all fixed.
